### PR TITLE
add file check_op_desc.py and add interface to get default value.

### PR DESCRIPTION
**Pull request by liym27**
> ### used for CI check https://github.com/PaddlePaddle/Paddle/pull/21522

## 1. add interface to get default value of attributes
```
m.def("get_op_attrs_default_value", [](py::bytes byte_name) 
           -> paddle::framework::AttributeMap{...}
```
---

## 2. add file check_op_desc.py to print op desc information and compare op desc
https://github.com/PaddlePaddle/Paddle/pull/21613 move get_all_ops_desc from check_op_desc.py to print_op_desc.py 

### (1) Print all ops desc in dict
~~run `python check_op_desc.py > OP_DESC_DEV.spec`  to get ops desc of branch develop~~
~~run `python check_op_desc.py > OP_DESC_PR.spec`  to get ops desc of branch PR~~

after #21613 :
run `python print_op_desc.py > OP_DESC_DEV.spec`  to get ops desc of branch develop~~
run `python print_op_desc.py > OP_DESC_PR.spec`  to get ops desc of branch PR~~

### (2) Compare op_desc files generated by branch develop and branch PR and print error message.
~~run `python check_op_desc.py OP_DESC_DEV.spec OP_DESC_PR.spec > error_message`~~

after #21613 :
run `python print_op_desc.py OP_DESC_DEV.spec OP_DESC_PR.spec`

---

- **for ops changed in branch PR:** 
  - add Input/Output but it's not dispensable
  - delele Input/Output
  - modify Input/Output including duplicable, intermediate or dispensable
  - add Attr but its default value is not set
  - delele Attr
  - modify Attr including the type, default value or  generated

- **for ops deleted in branch PR:** 
  - No comparison

- **for new ops only added in branch PR:** 
  - No comparison

### CODE REVIEW

File: tools/check_op_desc.py [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+
+    error_message = compare_op_desc(origin_op_desc, new_op_desc)
+    if error:
+        print_error_message(error_message)
+
+else:
+    print("Usage:\n" \
+          "\t1. python check_op_desc.py > OP_DESC_DEV.spec\n" \
+          "\t2. python check_op_desc.py > OP_DESC_PR.spec\n"\
+          "\t3. python check_op_desc.py > OP_DESC_DEV.spec OP_DESC_PR.spec > diff_message")
```
Comment (zhiqiu) [inline tools/check_op_desc.py]: Do you mean "python check_op_desc.py OP_DESC_DEV.spec OP_DESC_PR.spec > diff_message" ?

File: tools/check_op_desc.py [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+error = False
+
+
+def get_attr_default_value(op_name):
+    return core.get_op_attrs_default_value(cpt.to_bytes(op_name))
+
+
+def get_vars_info(op_vars_proto):
+    vars_info = {}
+    for vat_proto in op_vars_proto:
```
Comment (zhiqiu) [inline tools/check_op_desc.py]: There is a typo, `vat_proto` to `var_proto`.

Comment (liym27) [reply tools/check_op_desc.py]: Thx. done

File: paddle/fluid/pybind/pybind.cc [hunk 3]
```diff
@@ -1031,6 +1032,19 @@ All parameter, weight, gradient are variables in Paddle.
     }
     return ret_values;
   });
+  m.def("get_op_attrs_default_value",
+        [](py::bytes byte_name) -> paddle::framework::AttributeMap {
```
Comment (zhiqiu) [inline paddle/fluid/pybind/pybind.cc:1036]: Why not use `const std::string op_type` as input?

File: tools/check_op_desc.py (hunk 1 above)
Comment (liym27) [reply tools/check_op_desc.py]: Yes.  
Redundant `>` is deleted .

File: paddle/fluid/pybind/pybind.cc (hunk 3 above)
Comment (liym27) [reply paddle/fluid/pybind/pybind.cc:1036]: Both of `const std::string` and `py::bytes` can be use.

**Comment by zhiqiu** [comment]
LGTM.

File: python/paddle/fluid/tests/unittests/test_pybind_inference.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+import unittest
+from paddle.fluid import core
+from paddle import compat as cpt
+
+
+class TestPybindInference(unittest.TestCase):
+
+    # call get_op_attrs_default_value for c++ coverage rate
+    def test_get_op_attrs_default_value(self):
+        core.get_op_attrs_default_value(cpt.to_bytes("fc"))
```
Comment (luotao1) [inline python/paddle/fluid/tests/unittests/test_pybind_inference.py:25]: 可以在单测里，打印一下输出么？（下一个PR补上即可）

File: tools/check_op_desc.py [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+    Compare op_desc files generated by branch DEV and branch PR.
+    And print error message.
+    '''
+    with open(sys.argv[1], 'r') as f:
+        origin_op_desc = f.read()
+
+    with open(sys.argv[2], 'r') as f:
+        new_op_desc = f.read()
+
+    error_message = compare_op_desc(origin_op_desc, new_op_desc)
```
Comment (luotao1) [inline tools/check_op_desc.py:309]: 整体`check_op_desc`的速度（包括生成`OP_DESC_DEV.spec`和比较两个spec的区别）是多少？本地有验证过么？

可以在下一个PR上CI监控前验证下时间。

File: tools/check_op_desc.py [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+    if error:
+        print_error_message(error_message)
+
+else:
+    print("Usage:\n" \
+          "\t1. python check_op_desc.py > OP_DESC_DEV.spec\n" \
+          "\t2. python check_op_desc.py > OP_DESC_PR.spec\n"\
+          "\t3. python check_op_desc.py OP_DESC_DEV.spec OP_DESC_PR.spec > error_message")
```
Comment (luotao1) [inline tools/check_op_desc.py:317]: - 就用法来说，t1/t2是重复的。
- 建议分成两个文件来做，类似`tools/diff_api.py`和`tools/print_signatures.py`
- 报错需要告诉开发者复现途径，这里是否需要生成whl包后再判断呢？
https://github.com/PaddlePaddle/Paddle/blob/40c258a77b98c149d4df2b429cea5b12a8476a26/tools/diff_api.py#L27-L34

File: tools/check_op_desc.py [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+        attr_diff_message[DELETE] = attr_deleted_error_massage
+
+    return attr_error, attr_diff_message
+
+
+def compare_op_desc(origin_op_desc, new_op_desc):
+    origin = json.loads(origin_op_desc)
+    new = json.loads(new_op_desc)
+    error_message = {}
+    if cmp(origin_op_desc, new_op_desc) == SAME:
```
Comment (luotao1) [inline tools/check_op_desc.py:181]: 字符串比较的时间如何？是否要先生成md5值来比较

**Review by luotao1 — APPROVED** [review]
LGTM! 一些建议可以在上CI监控的PR里修改

File: tools/check_op_desc.py (hunk 5 above)
Comment (liym27) [reply tools/check_op_desc.py:309]: 我在本地生成一个OP_DESC_DEV.spec 的时间一般在2s左右；
比较2个spec的时间在2-3s.

File: python/paddle/fluid/tests/unittests/test_pybind_inference.py (hunk 4 above)
Comment (liym27) [reply python/paddle/fluid/tests/unittests/test_pybind_inference.py:25]: 可以的。谢谢

File: tools/check_op_desc.py (hunk 7 above)
Comment (liym27) [reply tools/check_op_desc.py:181]: 这里比较2个op desc的时间在1ms内，对效率影响不大，可以直接比较

File: tools/check_op_desc.py (hunk 6 above)
Comment (liym27) [reply tools/check_op_desc.py:317]: > 就用法来说，t1/t2是重复的。

感谢建议，这一点稍后将会简化。

> 建议分成两个文件来做，类似tools/diff_api.py和tools/print_signatures.py

感谢建议。之前确实是分成2个文件来做，但是考虑到两个文件很多共用的字符串，就统一在了一个文件中。
稍后将分成2个文件。

>报错需要告诉开发者复现途径，这里是否需要生成whl包后再判断呢？
 
需要生成whl包，稍后将加入复现途径。

**Pull request merged**